### PR TITLE
generalize fold_left; implement min_element

### PR DESCRIPTION
**Pull request by ecrypa**
Continuation of https://github.com/brunocodutra/metal/pull/99#discussion_r331799458.

I copied the code of `fold_left` because I want to benchmark old and new versions. Preliminary results can be found in the section "Change of fold left" at https://ecrypa.github.io/metal-benchmark/.

### CODE REVIEW

**Comment by ecrypa** [comment]
## Summary

This generalization of `fold_left` seems to work, but I do not like it. I will concisely try to explain my dislike. Then I turn back to `bind`-free composition _outside_ of `fold_left`. Finally, I ask specific questions.

## Why I dislike the approach of this PR

Semantically, a fold can be fully specified by a single `template<class init, class arg> class expr`. Instead, [this implementation (v2)](https://github.com/brunocodutra/metal/blob/0defd35334db8254395b9d1e3e1e536b3b088bd7/include/metal/value/fold_left_v2.hpp#L259) asks for a `comp` (which is an expression that specifies an ordering), and it additionally asks for that weird `template<template<class, class> class comp, class init, class arg> class combiner`. [Inside the fold](https://github.com/brunocodutra/metal/blob/0defd35334db8254395b9d1e3e1e536b3b088bd7/include/metal/value/fold_left_v2.hpp#L150), `comp` and `combiner` are composed to the real `expr` that is used for the folding. From my point of view, this is not a natural interface for `fold_left`, and other use cases may require even more arguments than just `comp` and `combiner`. Key question: **Why do we compose _inside_ of the `fold_left` implementation?**

## How to compose user-provided ordering and min_element logic

I can think of three options to consider:

1. composition inside `fold_left` (see above)
2. composition outside, based on `metal::bind` (works, but not blazingly fast; [see](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v10/) [v4](https://ecrypa.github.io/metal-benchmark/min_element_clang8_v4_v10/))
3. composition outside, based on native C++ templates ([SFINAE issues](https://github.com/ecrypa/metal/commit/11289ce3d1a3fad72c5de78c14fce40c1c61b65f#r35182663))

I believe that **(1)** diminishes the utility of `fold_left` as a general building block because the `fold_left` interface can offer a limited level of composition only.

Improving the performance of **(2)** is highly desirable because user-code performance can be expected to benefit from efficient composition, too. However, I expect the optimization to be difficult (impossible?) because it is already [better than (you @brunocodutra) expected](https://github.com/brunocodutra/metal/pull/99#pullrequestreview-297796402).

When [@brunocodutra pointed me at the SFINAE issues](https://github.com/ecrypa/metal/commit/11289ce3d1a3fad72c5de78c14fce40c1c61b65f#r35182663) of **(3)**, I was scared. I surely do not fully understand it yet, but I have gained some optimism that it can be mitigated. Therefor, I would like to understand **what kind of failures we have to consider**.

## Failures to be considered by approach (3) to `min_element`
... and how to handle them

### (A) user-provided ordering expression expects a wrong number of arguments
We should be able to preserve SFINAE-friendlyness by simple pattern matching.

### (B) user-provided ordering fires a `static_assert` on certain input
I believe that there is no way at all to preserve SFINAE-friendlyness in that case, right?

### (C) user-provided ordering (UPO) causes a SF when passed certain arguments
Since the arguments are [forwarded to the UPO](https://github.com/ecrypa/metal/commit/11289ce3d1a3fad72c5de78c14fce40c1c61b65f#diff-6da64bbd3d17cd953a6c08ef23524ab5R19) not before [`min_element` is "called"](https://github.com/ecrypa/metal/commit/11289ce3d1a3fad72c5de78c14fce40c1c61b65f#diff-6da64bbd3d17cd953a6c08ef23524ab5R29), there should be the desired SF on `min_element`.

### (D) ... (What else can happen?)
As far as I understand that whole SFINAE business, the following techniques should help in any conceivable case: We can change our external composition from
```
template<class init, class arg>
using combiner = if_<comp<arg, init>, arg, init>;
```
to
```
template<
  template<class> class di /* disguised_identity */,
  class init, class arg>
using combiner = if_<comp<di<arg>, di<init>>, di<arg>, di<init>>;
```
or
```
template<
  template<template<class...> class> dei /* disguised_expr_identity */,
  class init, class arg>
using combiner = if_<typename dei<comp>::template type<arg, init>, arg, init>;
```
or a combination of both.

If the private interface (for internal use only) of `fold_left` is changed such that it accepts one of the two latter aliases (or its combination) then we can achieve a much more general SFINAE-friendly composition than with the [current draft implementation](https://github.com/brunocodutra/metal/pull/100/files#diff-e9138a0774c8af01cfe400ed4a1b1964).

## Specific questions

I would be particularly thankful for input on any of these specific points:

- What do you think about the [internal composition **(1)** in this PR](https://github.com/brunocodutra/metal/pull/100/files#diff-e9138a0774c8af01cfe400ed4a1b1964)?
- Which failures other than (A)--(C) should I consider when testing native composition **(3)**?
- Do you agree that (D) should help in any case of failure? Can we get away without it, though?
- Any other hints how I should proceed?

**Comment by brunocodutra** [comment]
> Why do we compose inside of the fold_left implementation?

Agree 100% that this is not something we would like to do, considering Metal provides first class higher order composition that could just be reused. The only reason why we would want to do something like this would be to get better benchmarks, but frankly it may not be worth it, specially if we think we're paying for a performance improvement no one needs.

> However, I expect the optimization to be difficult (impossible?) because it is already better than (you @brunocodutra) expected.

A very pleasant surprise I should add :)

> We should be able to preserve SFINAE-friendlyness by simple pattern matching.

Unfortunately, [you can't do that in general](https://godbolt.org/z/xggnhN).

> I believe that there is no way at all to preserve SFINAE-friendlyness in that case, right?

Correct, more generally, any error that originates in the _definition_ as opposed to _declaration_ of a type  template is unrecoverable through SFINAE and considered a user error.

> Since the arguments are forwarded to the UPO not before min_element is "called", there should be the desired SF on min_element.

I'm not sure I understand this statement, could you elaborate?

> As far as I understand that whole SFINAE business, the following techniques should help in any conceivable case <snip>

I'm not sure I understand, how would `di` or `dei` help with SFINAE? 

Perhaps I should clarify what I mean by SFINAE-friendliness. 
I'll start by defining a way to testing for it:

```cpp
template<template<class...> class tmpl, class... args>
using is_sfinae_friendly = metal::is_invocable<metal::lambda<tmpl>, args...>;
```

For any `tmpl` and any `args`, `tmpl` is only SFINAE-friendly if instantiating `is_sfinae_friendly<tmpl, args...>` does not trigger a compilation error.

Metal unit tests systematically check for its own SFINAE-friendliness by attempting to evaluate algorithms on all sorts of [ill-defined types](https://github.com/brunocodutra/metal/blob/master/test/unit/include/test/values.hpp#L6). You'll see this test data has been carefully crafted to cover various edge cases and, in particular, attempts to trip `metal::as_number` by defining a nested `::value` to be various things other than the expected constant integral value convertible to `metal::int_`. You'll see that unit tests expect `metal::as_number` to [survive instantiations against all of these](https://github.com/brunocodutra/metal/blob/master/test/unit/src/metal/number/number.cpp#L6) without triggering a compilation error.

Checking whether some template expression is SFINAE-friendly is in general very complex and requires deep understanding of the standard that I do not claim to have myself. Put simply, a substitution failure is only recoverable if it's triggered as part of the _declaration_ of a type due to its own instantiation or the instantiation of one of its immediate type arguments. A substitution error that gets triggered as part of the _definition_ of a type is never recoverable. For the purpose of this definition, a template alias is considered a _declaration_.

If we turn back to the original definition of `_min_element_val`:

```
template<template<typename...> class expr>
struct _min_element_val<lambda<expr>> {
  template<typename x, typename y>
  using combiner = if_<expr<x, y>, x, y>;
  // ...
};
```

I claim `combiner` is not SFINAE friendly, because a substitution error that is triggered by `expr<x, y>` as part of the instantiation of `combiner<x, y>` though part of its declaration, is triggered by the substitution of `expr`, which is **not** one of its _immediate_ type arguments. Therefore such a substitution error leads to a compilation error. 

You can test this by attempting to instantiate `is_sfinae_friendly<_min_element_val<lambda<std::is_void>>, void, some_extra_argument>`, which should trigger a compilation error stating that `std::is_void<void, some_extra_argument>` cannot be instantiated.

I hope this helps.

File: include/metal/value/fold_left_v2.hpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    } // v2
+} // metal
+
+#include "../lambda/lambda.hpp"
+#include "../value/identity.hpp"
+
+#include <cstddef>
+
+namespace metal {
+    namespace v2 {
```
[the remark this answers is not in the document]
Comment (brunocodutra) [reply include/metal/value/fold_left_v2.hpp]: Not that anyone should care about a 7 year old compiler at this point, but I remember going [through great pains](https://github.com/brunocodutra/metal/issues/59#issuecomment-287603352) to support older GCC versions at some point, see #59.

**Comment by ecrypa** [comment]
Wow! Thanks a lot for your excellent explanations. I very much appreciate the time you are spending to teach me these things.

While your explanations are _extremely_ useful for me, I suspect that it is most efficient (for both of us) if I do not respond to each of your points individually. Instead, I propose to respond with (hopefully) improved code. Should you be more interested in my thoughts/explanations then please let me know.

**Comment by ecrypa** [comment]
I start to feel more confident about the application of these metaprogramming techniques. I am looking forward to your comments to this latest approach.

By the way: Should I close-and-open, or is it fine to continue here?

**Comment by ecrypa** [comment]
> > We should be able to preserve SFINAE-friendlyness by simple pattern matching.
> 
> Unfortunately, [you can't do that in general](https://godbolt.org/z/xggnhN).

What an [inconsistent behavior of compilers](https://godbolt.org/z/vAOJ9P) here! Clang matches nothing, MSVC matches everything, and GCC/ICC need C++17 in order to match how we want them to (C++14 is not enough). I can not tell if and where to submit bug reports.

**Comment by ecrypa** [comment]
[Benchmarks](https://ecrypa.github.io/metal-benchmark/) are promising. The winner [metal::v14::min_element](https://github.com/ecrypa/metal/blob/9fadd0c9ef2562ee0eb0f7deed63a7ef824c0af3/include/metal/list/min_element.v14.fast_dcall.hpp#L29) (similar to the version of this PR) is based on the new helper [dcall](https://github.com/ecrypa/metal/blob/9fadd0c9ef2562ee0eb0f7deed63a7ef824c0af3/include/metal/detail/sfinae.hpp#L79).

If there are no SFINAE troubles then this is my favorite version.

**Comment by brunocodutra** [comment]
>  I suspect that it is most efficient (for both of us) if I do not respond to each of your points individually

I'm fine either way, but I'll try to summarize all my thoughts and answer all your question on this single comment to hopefully make it easier for both of us to make sense of all the discussions we have in flight.

> Should I close-and-open, or is it fine to continue here?

It's totally fine to keep updating this one. If you're wondering, force-pushing is fine too.

> What an inconsistent behavior of compilers here!

Whoa, I didn't actually check other compilers. You know you've pushed too far when compilers stop agreeing with each other. I don't know which, if any, is correct here, but experience tells me clang is usually right.

> Benchmarks are promising (Clang 8, GCC 9). The winner `metal::v14::min_element` (similar to the version of this PR) is based on the new helper `dcall`.

Impressive! This may be it! Just wondering, what happens if we get rid of `class... _` in the definition of `combiner`?

I think `dcall` _may_ get even faster if you specialize `_dcaller` to help the compiler with memoization. I imagine something like this:

```cpp
template<bool>
struct _dcaller {};

template<>
struct _dcaller<true> {
    template<template<class...> class expr, class... vals>
    using type = expr<vals...>;
};

template<template<class...> class expr, class... vals>
using dcall = typename _dcaller<sizeof...(vals) >= 0>::template type<expr, vals...>;
```
Whether `dcall` is SFINAE-friendly, I don't know, but I quickly reimplemented `metal::bind` in terms of `dcall` instead of `_bind_impl` to see if it would work, but unfortunately it doesn't pass SFINAE test cases. I'd encourage you to define unit tests for `min_element` upfront to help filtering out alternatives that are not viable.

File: example/src/list.cpp [hunk 2]
```diff
@@ -677,4 +677,187 @@ IS_SAME(
...
+template<class, class>
+struct bad_comp {};
+
+static_assert(
+  !metal::is_invocable<
+  metal::lambda<metal::min_element>,
+  metal::numbers<4, 42, 17, -7, 0>,
+  metal::lambda<bad_comp>
+  >{}, "");
+)
```
Comment (brunocodutra) [inline example/src/list.cpp]: You probably realize this, but this file is intended to host examples that go in the doc, unit tests should be defined in therms of the preprocessor based suite.

**Comment by ecrypa** [comment]
> Just wondering, what happens if we get rid of `class... _` in the definition of `combiner`?

I had once introduced that pack in order to improve the performance by reducing the number of class-template instantiations. I wanted a single memoized instantiation of `_dcaller` to be reused for all `_dcall`s. At the same time, `_dcall` must provide SFINAE guarantees. The latter is achieved by making the `_dcaller` depend on the arguments. For performance reasons, I had tried to avoid `x` and `y` as template arguments of `_dcaller` because that causes many class-template instantiations. Instead, I introduced the argument-dependence by _disguising the actual number of arguments_. 

However, a benchmark has shown no significant differences in terms of compile-time. That is why [I have removed that empty pack](https://github.com/brunocodutra/metal/pull/100/commits/9f001e81e08ba023244cf882e9ea2ad444ce2bd4). The new version (current version of this PR) uses the plain old technique that is used all over the place in metal: The `_dcaller` depends on all arguments of the `expr`. One argument also worked, but I could not prove any significant compile-time advantage either.

I suspect that the memory footprint is better with the empty-pack trick, but I did not have any issues with lists of size 500+.

Without measurable advantage I classified that empty-pack trick as premature optimization.

> I think `dcall` _may_ get even faster if you specialize `_dcaller` to help the compiler with memoization. I imagine something like this:

Although I have discarded that empty-pack trick, I will keep that specialization technique in mind! Is it important to provide a definition of the unspecialized class template, or can I also omit the `{}`?

> Whether `dcall` is SFINAE-friendly, I don't know, but I quickly reimplemented `metal::bind` in terms of `dcall` instead of `_bind_impl` to see if it would work, but unfortunately it doesn't pass SFINAE test cases.

I observed that the `sizeof...` obfuscation does not work if the compiler sees the number of arguments too early. That is the reason why I had to introduce that empty pack in `combiner` (it was not sufficient to have a `_binary_dcall` with two arguments plus that empty pack).

Thanks for the hint to make use of existing unit tests.

> I'd encourage you to define unit tests for `min_element` upfront to help filtering out alternatives that are not viable.

When optimizing my approaches, I used some hand-crafted tests in the examples, which you also discovered. I did see plenty of other attempts failing, so this one should not be super unfriendly to SFINAE :-).

File: example/src/list.cpp (hunk 2 above)
Comment (ecrypa) [reply example/src/list.cpp]: Yes, you are right.

This is just more convenient to me, and I also wanted to show it to you. It is definitely not intented to be merged.

File: include/metal/detail/sfinae.hpp [hunk 3]
```diff
@@ -58,6 +58,15 @@ namespace metal {
...
 #endif
+
+        template<class... vals>
+        struct _dcaller {
+            template<template<class...> class expr>
+            using type = expr<vals...>;
+        };
+
+        template<template<class...> class expr, class... vals>
+        using dcall = typename _dcaller<vals...>::template type<expr>;
```
Comment (brunocodutra) [inline include/metal/detail/sfinae.hpp]: I wonder if you can just use `metal::invoke` instead of `detail::dcall` with no impact to performance. Notice that `metal::invoke` is defined in terms of `metal::apply`, which at some point [used to be implemented exactly this way](https://github.com/brunocodutra/metal/blob/630e8feb1da4388ba5c08b1e5f432c58c24b429a/include/metal/lambda/apply.hpp#L70). I [later changed it](https://github.com/brunocodutra/metal/commit/2071e36f3bbcf6fbe57f94a568f34812ad581c7e#diff-5f827d47769fbe9312e9e904c05fbc3eR53) along with the introduction of `metal::cascade`, but I can't remember why, but likely due to some compiler that didn't want to cooperate. I don't know whether that change had any impact in benchmarks, but we could definitely attempt to revert it back to the previous implementation to avoid introducing `dcall`.

**Comment by ecrypa** [comment]
> Without measurable advantage I classified that empty-pack trick as premature optimization.

Just to back this up: [here](https://ecrypa.github.io/metal-benchmark/min_element_clang9_type_count/time.html) are the benchmarks using Clang 9 (see v11, v12, v13, and v14 in particular). Meanwhile, I have [customized metabench to measure the "number of types"](https://github.com/ecrypa/metabench/commit/6578fc6b5d43b63b2efd4fc5af23158b03133afe) as reported by Clang with `-Xclang -print-stats`. [Results do differ](https://ecrypa.github.io/metal-benchmark/min_element_clang9_type_count/), but there are no major differences among v11..v14. I should add that I do not understand the output of `-Xclang -print-stats` (see sample [here](https://ecrypa.github.io/metal-benchmark/sample_clang_stats_output.txt)), and I just used the number from the second line in this quote:

> *** AST Context Stats:
>  73 types total.

**Comment by brunocodutra** [comment]
Amazing work! Do consider contributing that to `metaben.ch`, it would be awesome to have a toggle button to switch between time and number of types.

**Comment by ecrypa** [comment]
> Do consider contributing that to `metaben.ch`,

I have opened an issue there in order to share that idea.

File: include/metal/detail/sfinae.hpp (hunk 3 above)
Comment (ecrypa) [reply include/metal/detail/sfinae.hpp]: > I wonder if you can just use `metal::invoke` instead of `detail::dcall` with no impact to performance. 

Thanks for the suggestion. It [does](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v15_bestof/) [work](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v15_bestof/) (see [v15](https://github.com/brunocodutra/metal/commit/22844ee3eb780a8539a707a1fc088c839772795a)).

I had not expected that `invoke` is implemented in this way. I was assuming an alias.

> Notice that `metal::invoke` is defined in terms of `metal::apply`, which at some point [used to be implemented exactly this way](https://github.com/brunocodutra/metal/blob/630e8feb1da4388ba5c08b1e5f432c58c24b429a/include/metal/lambda/apply.hpp#L70). I later [changed it](https://github.com/brunocodutra/metal/commit/2071e36f3bbcf6fbe57f94a568f34812ad581c7e#diff-5f827d47769fbe9312e9e904c05fbc3eR53) along with the introduction of `metal::cascade`, but I can't remember why, most likely some compiler didn't want to cooperate.

I have made the following observation: Primary class templates with huge packs are expensive to instantiate. I can imagine that specialization (pattern matching `list<vals...>`) is more efficient.

> I don't know whether that change had any impact in benchmarks, but we could definitely attempt to revert it back to the previous implementation to avoid introducing `dcall`.

We can use `invoke` just fine.

**Comment by ecrypa** [comment]
I like how this is boiling down to a two-liner (after some [crazy](https://github.com/ecrypa/metal/blob/22844ee3eb780a8539a707a1fc088c839772795a/include/metal/list/min_element.v7.alias_fasttrack.hpp#L20) [intermediate](https://github.com/ecrypa/metal/blob/22844ee3eb780a8539a707a1fc088c839772795a/include/metal/list/min_element.v8.foldlike.hpp#L46) [attempts](https://github.com/ecrypa/metal/blob/22844ee3eb780a8539a707a1fc088c839772795a/include/metal/list/min_element.v10.foldlike_indirection.hpp#L46)).

Latest benchmarks of selected versions:
- [Clang 9](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v15_bestof/)
- [GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v15_bestof/)

**Comment by brunocodutra** [comment]
I missed one of your questions last time.

> Is it important to provide a definition of the unspecialized class template, or can I also omit the {}?

Without the `{}`, types are declared but not defined and I fear that may open a whole can of worms when it comes to ODR violations and undefined behavior, so I prefer to define everything.

**Comment by brunocodutra** [comment]
Benchmarks look pretty sweet! Those extra 30 ms on a list with 1000 elements definitely don't seem worth the pile of code that fast tracking requires.

I wonder if redefining `metal::apply` in a similar way to `dcall` would shave off some of those 30 ms, but either way that's definitely out of scope for this PR.

**Comment by ecrypa** [comment]
> Benchmarks look pretty sweet! Those extra 30 ms on a list with 1000 elements definitely don't seem worth the pile of code that fast tracking requires.

The fastest solutions are the neat ones. One exception is the bulky solution v7.alias_fasttrack, which is fastest on clang, but slowest on GCC.

> I wonder if redefining `metal::apply` in a similar way to `dcall` would shave off some of those 30 ms, but either way that's definitely out of scope for this PR.

By that kind of optimization we can tweak the efficiency in the range of v11...v15. Fastest (and hardest to use) are v11 and v14, but that advantage is small and Clang-only. GCC 9 is equally fast with v11..v15.

**Comment by ecrypa** [comment]
I just benchmarked two different approaches to obtain the index (instead of the value). The winner (currently in this PR) is [v17.ind.find](https://github.com/ecrypa/metal/blob/76b7e92beba803838fe440a35ebf3fdc3bc65d89/include/metal/list/min_element.v17.ind.find.hpp#L17). That was surprising to me, because my favorite was [v16.ind.accumulate](https://github.com/ecrypa/metal/blob/76b7e92beba803838fe440a35ebf3fdc3bc65d89/include/metal/list/min_element.v16.ind.accumulate.hpp#L15). I should analyze if the huge difference originates from `indices<seq>` or from the `accumulate` itself.

See benchmark results for [Clang 9](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v17_bestof/) and [GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v17_bestof/).

Other ideas how to obtain the index?

**Comment by brunocodutra** [comment]
> Other ideas how to obtain the index?

Have you tried `metal::transpose`?

```.cpp
template<class lbd, class is_empty>
struct _min_element {};

template<class lbd>
struct _min_element<lbd, true_> {
    template<class>
    using type = number<0>;
};

template<class lbd>
struct _min_element<lbd, false_> {
    template<class state, class next>
    using custom_min = if_<apply<lbd, back<transpose<next, state>>>, next, state>;

    template<class seq>
    using type = front<apply<lambda<custom_min>, transpose<pair<indices<seq>, seq>>>>;
};
```

**Note:** not tested.

**Comment by ecrypa** [comment]
For huge lists, it is slower than `find` ([Clang 9](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v19_messy/), [GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v19_messy/)). [v19](https://github.com/ecrypa/metal/blob/0962b6e008306a51f44df4a06cae688600e3a186/include/metal/list/min_element.v19.ind.transpose.hpp) is very close to your version, [v18](https://github.com/ecrypa/metal/blob/0962b6e008306a51f44df4a06cae688600e3a186/include/metal/list/min_element.v18.ind.transpose.hpp) is slightly optimized (but also slower than `find` for huge lists). Smaller lists are hard to measure.

**Comment by brunocodutra** [comment]
Can we stack this against a version that returns the value itself, so we can see the cost of recovering the index?

**Comment by ecrypa** [comment]
You can estimate the cost of recovering the index from the latest benchmarks. v11 is the fastest version, and it returns the value. v15 is my most favorite version, and it returns the value, too. Index return is implemented in v16.ind, v17.ind, v18.ind, and v19.ind only. The latter (v16..v19) are based on v15 (using `invoke` or `apply`). You can find the implementations in my fork's branch called [`benchmark/min-element`](https://github.com/ecrypa/metal/tree/benchmark/min_element/include/metal/list); permalink to the current version is [here](https://github.com/ecrypa/metal/tree/0962b6e008306a51f44df4a06cae688600e3a186/include/metal/list).

Do we need better resolution for small lists?

**Comment by brunocodutra** [comment]
I missed the fact v11 was returning the value, thanks.

> Do we need better resolution for small lists?

I don't think it's worth it, small lists are certainly not the bottleneck of a metaprogram.

**Comment by ecrypa** [comment]
TODO:

- [ ] final decision on interface
- [ ] final decision on implementation
- [ ] complement examples
- [ ] add tests
- [ ] pass code review

File: test/unit/src/metal/list/min_element.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    CHECK((metal::is_invocable<metal::lambda<metal::min_element>, LAMBDA(_), LAMBDA(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::min_element>, LAMBDA(_), LAMBDA(_)>), (FALSE)); \
+    CHECK((metal::min_element<metal::list<NUMBERSX20(INC(M))>, metal::lambda<metal::greater>>), (metal::number<M>)); \
+    CHECK((metal::min_element<metal::list<NUMBERSX20(INC(M))>, metal::lambda<metal::less>>), (metal::number<0>)); \
+    CHECK((metal::min_element<metal::list<NUMBERSX20(INC(M))>>), (metal::number<0>)); \
+    CHECK((metal::min_element<metal::list<ENUM(INC(N), NUMBERS FIX(INC(M)))>, metal::lambda<metal::less>>), (metal::number<0>)); \
+    CHECK((metal::min_element<metal::list<ENUM(INC(N), NUMBERS FIX(INC(M)))>>), (metal::number<0>)); \
+/**/
+
+GEN(MATRIX)
```
[the remark this answers is not in the document]
Comment (ecrypa) [reply test/unit/src/metal/list/min_element.cpp]: The test is derived from the test of `metal::sort`.

It bothers me that the [test is passing](https://travis-ci.org/brunocodutra/metal/jobs/603225564#L536) while the [example fails](https://travis-ci.org/brunocodutra/metal/jobs/603225564#L415)  on GCC 4.7.

File: include/metal/list/min_element.hpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    /// \tip{`lbd` may be omitted, in which case it defaults to `metal::lambda<metal::less>`.}
+    ///
+    /// ### Example
+    /// \snippet list.cpp min_element
+    ///
+    /// ### See Also
+    /// \see min, sort
+#if !defined(METAL_WORKAROUND)
+    template<class seq, class lbd = metal::lambda<metal::less>>
+    using min_element = apply<
```
[the remark this answers is not in the document]
Comment (ecrypa) [reply include/metal/list/min_element.hpp]: Testing your suggestion in #112. As an alternative, I am also testing 9d833d89119cf306ff0f0612056ef3dc9d1ed6d1 in this PR.

**Comment by ecrypa** [comment]
Good news is that 39a5465 passes the [list example on GCC 4.7](https://ci.appveyor.com/project/brunocodutra/metal/builds/28439214/job/kpg8uglcs52r06v4#L67). Bad news is that it (v22) is [slower than higher-order metafunction (v4) on GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v22_gcc_fix/). Weirdly, [Clang 9](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v22_gcc_fix/) is fastest with this fix.

To be improved...

**Comment by ecrypa** [comment]
Nevermind, I was looking at MSVC out of habit. [GCC 4.7 still fails](https://travis-ci.org/brunocodutra/metal/jobs/604082788#L415) to compile the example :-1: .

**Comment by ecrypa** [comment]
For some reason, GCC 4.7 chooses the overload `value<> sfinae(...);` for the `min_element` example. I turned it into a hard error and reduced it to a [minimal example](https://godbolt.org/z/aT9ipi).

Can you give me any advices how to overcome that issue?

```C++
#define WORKAROUND_THAT_BREAKS_GCC_4_7
// removing the line above makes GCC 4.7 happy

template<
  template<template<class...> class...> class tmpl,
  template<class...> class... exprs>
struct _forwarder {
  using type = tmpl<exprs...>;
};

template<
  template<template<class...> class...> class tmpl,
  template<class...> class... exprs>
#if defined(WORKAROUND_THAT_BREAKS_GCC_4_7)
using forward = typename _forwarder<tmpl, exprs...>::type;
#else
using forward = tmpl<exprs...>;
#endif

struct _min_element_impl {
  template<template<class...> class>
  using type = long;
};

template<template<class...> class expr>
struct _min_element {
  using type = forward<_min_element_impl::template type, expr>;
};

template<class T>
using some_expr = T;

using some_val = _min_element<some_expr>::type;

some_val foo = 4;
```

**Comment by brunocodutra** [comment]
If GCC 4.7 is causing so much trouble, it may be time we let it go.
All it takes is [updating Travis configuration](https://github.com/brunocodutra/metal/blob/c17c8ad6d65e07b2eb2903f0f404bc5792a0c28c/.travis.yml#L29-L41).
